drivers: spi_sifive: Optimize xfer to minimize downtime between frames.

snematbakhsh opened the pull request:
The SPI peripheral on SiFive parts uses FIFOs for Rx and Tx (FIFO size
undocumented, but empirically found to be 8 bytes on FE310, likely
identical on FU540 / FU740). Make use of these FIFOs in order to
continuiously feed Tx data as available.

Verified to transmit 1 MHz SPI @ 200 MHz coreclk / tlclk on FE310
continuously without downtime between frames.

Signed-off-by: Shawn Nematbakhsh <shawn@rivosinc.com>

carlescufi commented:
@tbursztyka, @henrikbrixandersen and @anangl I know this is not a platform you use, but I thought that, as a general SPI optimization maybe you'd be able to take a look

katsuster reviewed (dismissed):
LGTM. And I can use SPI1 successfully with the patch on my HiFive1 rev.B board.

Tested-by: Katsuhiro Suzuki <katsuhiro@katsuster.net>

Current HiFive1 rev.B SPI config is broken. If you want to test on HiFive1 please add my patch (PR #45651).

henrikbrixandersen reviewed (dismissed):
LGTM

anangl commented on drivers/spi/spi_sifive.c:
```diff
@@ -99,50 +100,63 @@ int spi_config(const struct device *dev, uint32_t frequency,
 
-	while ((val = sys_read32(SPI_REG(dev, REG_RXDATA))) & SF_RXDATA_EMPTY) {
+	if (reg & SF_RXDATA_EMPTY) {
+		return false;
 	}
-
-	return (uint16_t) val;
+	*val = reg;
```
It's better to keep this explicit conversion from `uint32_t` to `uint16_t`, otherwise compilers may issue a warning for this, depending on the configuration used (see `-Wconversion` option in gcc, for example).

snematbakhsh replied:
Fixed.

anangl commented on drivers/spi/spi_sifive.c:
```diff
@@ -99,50 +100,63 @@ int spi_config(const struct device *dev, uint32_t frequency,
-	} while (spi_context_tx_on(ctx) || spi_context_rx_on(ctx));
+		if (queued_frames > 0 && spi_sifive_recv(dev, &rxd)) {
+			if (spi_context_rx_buf_on(ctx)) {
+				*ctx->rx_buf = rxd;
```
Also here explicit conversion should be used (`ctx->rx_buf` points to an `unit8_t`).
And why `rxd` and `txd` are of type `uint16_t`, actually (only 8 lower bits of both are used in fact)?

snematbakhsh replied:
> And why rxd and txd are of type uint16_t, actually (only 8 lower bits of both are used in fact)?

You are right, uint8_t seems appropriate. I added a separate CL to this PR to change these to uint8_t.

snematbakhsh committed:
drivers: spi_sifive: Use uint8_t data/buffer types to match spi_context.

spi_context uses uint8_t buffers, and HW rxdata / txdata registers only
contain 8 bits of data (along with b31 full / empty flag), so uint8_t is
appropriate.

Suggested-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
Signed-off-by: Shawn Nematbakhsh <shawn@rivosinc.com>

snematbakhsh committed:
drivers: spi_sifive: Optimize xfer to minimize downtime between frames.

The SPI peripheral on SiFive parts uses FIFOs for Rx and Tx (FIFO size
undocumented, but empirically found to be 8 bytes on FE310, likely
identical on FU540 / FU740). Make use of these FIFOs in order to
continuiously feed Tx data as available.

Verified to transmit 1 MHz SPI @ 200 MHz coreclk / tlclk on FE310
continuously without downtime between frames.

Signed-off-by: Shawn Nematbakhsh <shawn@rivosinc.com>

snematbakhsh commented on drivers/spi/spi_sifive.c:
```diff
@@ -99,50 +100,63 @@ int spi_config(const struct device *dev, uint32_t frequency,
 	return 0;
 }
 
+static ALWAYS_INLINE bool spi_sifive_send_available(const struct device *dev) {
```
Thanks for the review. Fixed.
